### DISCUSSION TITLE
Current development plans + Gyromina v2.0

### DISCUSSION THREAD

**Discussion by Lowie375** [Lowie375/Gyromina#99] [Announcements]
### I have a rough idea of where I want Gyromina to be in the near future, so I thought I'd share it.

As I stated in my last announcement (#84), implementing slash command support is my main priority right now. Beyond that, however, I have a few plans for updates post-slash command migration, including one major one that will likely lead to the development of Gyromina v2.0.

Listed below is an overview of my current prospective roadmap, along with a few other comment about future updates to Gyromina.

#### v1.3 - Slash Command Migration + QoL (#95)

This is what I'm currently working on, and is my first priority going forward, since this *could* become time-sensitive at some point with Discord's move away from message commands in favour of slash command interactions. I'd rather be safe than sorry. 

- Migrate existing commands to slash-compatible frameworks
- Add new functionality for message-reliant commands
- Restructure repository framework

#### v1.4 - Game Saves

The ability to save games is something I've been considering adding for a while, mainly because I've had experiences where I've executed time-sensitive actions and lost my progress on them because I was distracted or interrupted. Adding a way to save a game (or command) and be able to come back to it later would provide a way to get around this issue, so it's something I wanted to add as another quality of life feature.

- Restructure game files into initialization + core functionality
- Add save + load (re-initialization) systems

#### v1.5 - Multiplayer Games

This is something I've been wanting to do for quite some time now, but have never gotten around to (due to a lack of time and the prioritization of other features). 

- Add framework for multiplayer games
- Create special save state options for individual player data + open board data

#### v2.0 (the big one) - Abstraction + Modularization

**TL;DR: I want to pack parts of Gyromina's command functionality into an `npm` module for external use.**

The idea for this came about when I began development for one of my other projects, [prontbot](https://github.com/Lowie375/prontbot), and realized that I wanted it and Gyromina to share some parts of their functionality (notably, the `roll` command). Obviously, this could be done by copy-pasting from Gyromina into prontbot and refactoring it to work with `tmi.js` rather than `discord.js`, but that would then make bugfixing twice as hard each time I found an issue with the copied code.  
I needed a way to access some of Gyromina's command functionality from another scope... so what better way to do to that than to ship parts of Gyromina's functionality in an `npm` module!

My current plan is to abstract non-`discord.js`-specific code into new, exportable functions and make those functions accessible through the `gyromina` module.  
Obviously, there will be quite a few challenges that will need to be overcome to make this a reality, such as figuring out how to not make a Gyromina run as a bot by default when imported, how to format abstracted commands and properly reintegrate them as bot commands, how to format exports, and a whole lot more.

This will be a big undertaking, and will likely result in a massive rewrite of the majority of Gyromina's code, hence why it'll likely be released as Gyromina v2.0. Additionally, this will take a lot of time to properly complete, so it'll likely be developed alongside some other minor updates and implemented much farther into the future. As always, help is appreciated, so if you're willing to offer any tips on how to make this happen or assistance in doing so, I'd be eternally grateful.

#### So, yeah, that's the roadmap!

All these updates will take me some time to complete, and I've only just begun work on v1.3, so don't expect any of these updates to be completed incredibly soon. Development takes time, and I know this roadmap will keep me busy for quite a lot of it.  
It's also worth noting that none of these updates (bar v1.3) are *completely* set it stone, so they may be subject to change. Who knows, maybe another new idea will end up delaying multiplayer games even further, or maybe v2.0 will come sooner than expected!

Regardless, this is where I'm currently at, and I'm excited for the future of Gyromina! Here's to continued development for the years to come!
